galaxy_importer/loaders.py

```diff
@@ -140,52 +139,66 @@ def _make_path_name(rel_path, name):
 
 
 class DocStringLoader():
-    def __init__(self, content_type, path, fq_name, logger=None):
-        self.content_type = content_type
+    """Process ansible-doc doc strings for entire collection.
+
+    Load by calling ansible-doc once in batch for each plugin type."""
+    def __init__(self, path, fq_collection_name, logger=None):
         self.path = path
-        self.fq_name = fq_name
+        self.fq_collection_name = fq_collection_name
         self.log = logger or default_logger
 
     def load(self):
-        if self.content_type not in ANSIBLE_DOC_SUPPORTED_TYPES:
-            return None
-
         self.log.info('Getting doc strings via ansible-doc')
-        json_output = self._run_ansible_doc()
-
-        if not json_output:
-            return None
-
-        data = json.loads(json_output)
-        if not isinstance(data, dict):
-            self.log.error('ansible-doc output not dictionary as expected')
-            return None
-        if len(data.keys()) != 1:
-            self.log.error('ansible-doc output did not return single top-level key')
-            return None
-        data = list(data.values())[0]
-        data = self._transform_doc_strings(data)
-
-        return {
-            key: data.get(key, None)
-            for key in ANSIBLE_DOC_KEYS
-        }
-
-    def _run_ansible_doc(self):
+        docs = {}
+        for plugin_type in ANSIBLE_DOC_SUPPORTED_TYPES:
+            plugin_dir_name = ANSIBLE_DOC_PLUGIN_MAP.get(plugin_type, plugin_type)
+
+            plugins = self._get_plugins(os.path.join(self.path, 'plugins', plugin_dir_name))
+
+            if plugins:
+                data = self._run_ansible_doc(plugin_type, plugins)
+                data = self._process_doc_strings(data)
+                docs[plugin_type] = data
+
+        return docs
+
+    def _get_plugins(self, plugin_dir):
+        plugins = []
+        for root, _, files in os.walk(plugin_dir):
+            for filename in files:
+                if not filename.endswith('.py') or filename == '__init__.py':
+                    continue
+                file_path = os.path.join(root, filename)
+                sub_dirs = os.path.relpath(root, plugin_dir)
+
+                fq_name_parts = [self.fq_collection_name]
+                if sub_dirs and sub_dirs != '.':
+                    fq_name_parts.extend(sub_dirs.split('/'))
+                fq_name_parts.append(os.path.basename(file_path)[:-3])
```

Comment:
Where's the '3' in `[:-3]` come from?  To remove '.py' from filename?

Reply:
that's right